crates/duty-tracker/src/s2_session_manager.rs

```diff
@@ -0,0 +1,184 @@
+            .new_session(
+                outpoint,
+                self.operator_table
+                    .btc_keys()
+                    .into_iter()
+                    .map(|x| x.to_x_only_pubkey())
+                    .collect(),
+                taproot_witness,
+                outpoint.txid,
+                outpoint.vout,
+            )
+            .await?
+            .map_err(MusigSessionErr::SecretServiceNewSessionErr)?;
+
+        let ours = first_round.our_nonce().await?;
+
+        self.first_round_map
+            .lock()
+            .await
+            .insert(outpoint, first_round);
+
+        Ok(ours)
+    }
+
+    /// Load a PubNonce into the signing session identified by OutPoint.
+    pub async fn put_nonce(
+        &self,
+        outpoint: OutPoint,
+        sender: XOnlyPublicKey,
+        nonce: PubNonce,
+    ) -> Result<(), MusigSessionErr> {
+        let mut guard = self.first_round_map.lock().await;
+        let first_round = guard.get_mut(&outpoint).ok_or(MusigSessionErr::NotFound)?;
+        Ok(first_round.receive_pub_nonce(sender, nonce).await??)
+    }
+
+    /// Given an OutPoint and the sighash for what is being signed get our PartialSignature.
+    pub async fn get_partial(
+        &self,
+        outpoint: OutPoint,
+        sighash: Message,
+    ) -> Result<PartialSignature, MusigSessionErr> {
+        if let Some(second_round) = self.second_round_map.lock().await.get_mut(&outpoint) {
+            Ok(second_round.our_signature().await?)
+        } else {
+            let mut first_guard = self.first_round_map.lock().await;
+            if let Some(first_round) = first_guard.remove(&outpoint) {
+                if first_round.holdouts().await?.is_empty() {
+                    let second_round = first_round.finalize(*sighash.as_ref()).await??;
+                    drop(first_guard);
+                    let ours = second_round.our_signature().await?;
+                    let mut second_guard = self.second_round_map.lock().await;
+                    second_guard.insert(outpoint, second_round);
+                    Ok(ours)
+                } else {
+                    first_guard.insert(outpoint, first_round);
+                    drop(first_guard);
+                    Err(MusigSessionErr::Premature)
+                }
+            } else {
+                Err(MusigSessionErr::NotFound)
+            }
+        }
+    }
+
+    /// Load a PartialSignature into the signing session identified by OutPoint.
+    pub async fn put_partial(
+        &self,
+        outpoint: OutPoint,
+        sender: XOnlyPublicKey,
+        partial: PartialSignature,
+    ) -> Result<(), MusigSessionErr> {
+        let mut guard = self.second_round_map.lock().await;
+        let second_round = guard.get_mut(&outpoint).ok_or(MusigSessionErr::NotFound)?;
+        Ok(second_round.receive_signature(sender, partial).await??)
+    }
+
+    /// Finalize the musig signing process and extract the final LiftedSignature
```

Comment:
```suggestion
    /// Finalizes the MuSig2 signing process and extract the final [`LiftedSignature`].
```

Amazing how we are FP'ing the codebase with terms 😄